### PR TITLE
--read-write-tmpfs=false should set /dev/* tmpfs to readonly

### PR DESCRIPTION
**Pull request by rhatdan**
Change the --read-only-tmpfs option to --read-write-tmpfs since
this makes sense and the old name was doing the exact opposite.

Fixes: #12937

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@giuseppe When I do this I attempt to make /dev, /dev/tty, /dev/shm, /dev/mqueue as read-only but crun blows up with 

./bin/podman run -ti --read-only --read-only-tmpfs=false alpine sh
Error: OCI runtime error: crun: opening file `/dev/console` for writing: Read-only file system

**Comment by giuseppe** [comment]
`/dev` is still used by the OCI runtime after setting the mounts.

I've fixed an issue in crun so it works when `--tty` is not used, but for a read-only `/dev` also in the tty case, we'd need to set the `ro` flag later.  I'll take a look at it.

I've tried with runc and it fails as well, so I think it should be optional to set `/dev` read-only.  Could we extend `--read-only-tmpfs` to accept a string and have something like `--read-only-tmpfs=all`?

**Comment by giuseppe** [comment]
this seems to help with your PR: https://github.com/containers/crun/pull/859

**Comment by rhatdan** [comment]
@kolyshkin  WDYT?

**Comment by rhatdan** [comment]
I think if we leave any tmpfs writable, we don't solve the problem.

**Comment by rhatdan** [comment]
Fixes: https://github.com/containers/podman/issues/12937

**Comment by kolyshkin** [comment]
As @giuseppe mentioned earlier, runc can't work with ro /dev, but this fix is easy https://github.com/opencontainers/runc/pull/3345 (and very similar to https://github.com/containers/crun/pull/857).

One other thing, @rhatdan, I also find it very confusing how `--read-only-tmpfs=false` results in read-only `/dev` mounts. Maybe a separate option (`--read-only-dev`) or something else entirely?

**Comment by rhatdan** [comment]
@kolyshkin I am not crazy about adding another function, Since I believe the original goal of --read-only-tmpfs (A badly named option, which we should probably change) was to make the entire container read-only. (At least the way I think about it).

**Comment by rhatdan** [comment]
We should change the option to read-write-tmpfs.

**Comment by rhatdan** [comment]
This PR is waiting for an updated version of runc, which will allow it to work with /dev set to read/only mode.

**Comment by kolyshkin** [comment]
> This PR is waiting for an updated released version of runc, which will allow it to work with /dev set to read/only mode.
> @kolyshkin @AkihiroSuda Any idea when that will be?

Depending on how soon you need it.

We can certainly release runc 1.1.1, either now (= soon) with this change alone, or a bit later, waiting for a few more fixes to pile up.

**Comment by kolyshkin** [comment]
> This PR is waiting for an updated released version of runc,

Backport PR: https://github.com/opencontainers/runc/pull/3355

**Comment by rhatdan** [comment]
Well since this is a post 4.0 feature, it is not needed immediately, but soon.

**Comment by rhatdan** [comment]
@cevich looks like ubuntu does not have the latest runc in it?

File: .cirrus.yml [hunk 1]
```diff
@@ -38,7 +38,7 @@ env:
     UBUNTU_NAME: "ubuntu-2110"
 
     # Google-cloud VM Images
-    IMAGE_SUFFIX: "c5814666029957120"
+    IMAGE_SUFFIX: "c5309576894480384"
```
Comment (cevich) [inline .cirrus.yml:41]: Can you use `c6118659411148800` please?

Comment (cevich) [reply .cirrus.yml:41]: oops, nvm, we'll need new images again for `runc` :cry:

**Comment by cevich** [comment]
> @cevich looks like ubuntu does not have the latest runc in it?

@rhatdan any hint what version or commit you need?

@lsm5 would you mind building new `crio-runc` in testing kubik?

**Comment by haircommander** [comment]
I'd rather rebuild when the 1.1.1 release is fully cut, which it doesn't seem to be yet (unless I'm missing something)

**Comment by rhatdan** [comment]
@haircommander Any progress?

**Comment by haircommander** [comment]
It doesn't look like runc 1.1.1 is release yet

**Comment by rhatdan** [comment]
@kolyshkin Estimated release date?

**Comment by kolyshkin** [comment]
> @kolyshkin Estimated release date?

This week (fingers crossed), hopefully today or tomorrow. Preparing the changelog now. See https://github.com/opencontainers/runc/issues/3437 to track progress.

**Comment by kolyshkin** [comment]
> > @kolyshkin Estimated release date?
> 
> This week (fingers crossed), hopefully today or tomorrow. Preparing the changelog now. See [opencontainers/runc#3437](https://github.com/opencontainers/runc/issues/3437) to track progress.

runc 1.1.1 is now released (it is amazing what can be done while being on vacation; perhaps we should all be on vacation permanently). https://github.com/opencontainers/runc/releases/tag/v1.1.1

**Comment by rhatdan** [comment]
@haircommander Does this mean you have something to do now, or is this in @cevich box?

**Comment by cevich** [comment]
lol, probably in my box now :smile:

**Comment by cevich** [comment]
Update: Nevermind, we need new images anyway for https://github.com/containers/storage/pull/1176

**Comment by cevich** [comment]
Images built fine, but I just checked and F35 only has runc-1.1.0-1.fc35.x86_64, and we're using the `updates-testing` repo.  :disappointed:   Can somebody press the magic buttons to get this packaged for F35 & 36?

**Comment by cevich** [comment]
I just noticed v1.1.1 is now in updates-testing, so I'm re-spinning VM images in https://github.com/containers/automation_images/pull/119 that should pick it up.

**Comment by cevich** [comment]
...Alright, give `c5706681618268160` a try.

**Comment by cevich** [comment]
***Note***: This is slightly newer, and contains a fix for multi-arch image building (don't clobber #13739): `c6464310661611520`

**Comment by cevich** [comment]
Clarification: I'd like to avoid the situation where #13739 merges, then this merges (with the old image ID), wrecking one of the multi-arch build task fixes (the fix is in the VM images).

**Comment by giuseppe** [comment]
it seem on ubuntu there is a too old crun version.  The error was fixed in crun-1.4.2

**Comment by umohnani8** [comment]
@rhatdan is this still on track for 4.1, it is looking difficult now. Removing the 4.1 label from it, but please add it back if you think it will make it in time.

**Comment by rhatdan** [comment]
@edsantiago Please give me a clue as to what is happening here on upgrade tests?

**Comment by edsantiago** [comment]
I have no idea. The only instance in which I see pod + freeing + lock is #8279, but that was a flake and got closed a year and a half ago.  This one is consistent.

For the record, the error here is:
```
# # podman pod rm mypod
# time="2022-05-13T05:00:08-05:00" level=error msg="Freeing pod dea923c9c17bc955a4552295d4cd2f58cee71c21e3442b2bbf066d7972ded896 lock: no such file or directory"
# Error: error freeing lock for container ca3fd43a4a31e6a6b9746cde748023c0e967fea53961e82e095a2c81511521e8: no such file or directory
```
My guess is that, with $TMPDIR being ro, something in the running podman processes is croaking. Those logs are not easy to find, so I'll have to take a look on Monday.

File: cmd/podman/common/create.go [hunk 2]
```diff
@@ -406,8 +406,8 @@ func DefineCreateFlags(cmd *cobra.Command, cf *entities.ContainerCreateOptions,
 			"Make containers root filesystem read-only",
 		)
 		createFlags.BoolVar(
-			&cf.ReadOnlyTmpFS,
-			"read-only-tmpfs", true,
+			&cf.ReadWriteTmpFS,
+			"read-write-tmpfs", true,
```
Comment (edsantiago) [inline cmd/podman/common/create.go:410]: So, I just gave this a quick look, and I'm a little confused: to my untrained eye, if a flag name changes from `read-only` to `read-write`, shouldn't its default change from `true` to `false`?

**Comment by vrothberg** [comment]
Friendly ping. @rhatdan, are you on it

**Comment by rhatdan** [comment]
I look at it once in a while and restart it hoping for a miracle.  But I have no idea why this does not work on Ubuntu.

**Comment by cevich** [comment]
Hrmmm, you're using recent VM images too :confused:  I see this, which is probably telling:

`runc: runc create failed: unable to start container process: unable to setup user: fchown fd 0: read-only file system: OCI runtime error`

Have you tried running any of those tests or failing commands manually under `hack/get_ci_vm.sh`?  Might help get to the bottom of why/how the filesystems are going/set read-only.

**Comment by rhatdan** [comment]
Nope, Have thought about it, but never have time.  Others can take if over and drive to conclusion...

**Comment by kolyshkin** [comment]
> runc: runc create failed: unable to start container process: unable to setup user: fchown fd 0: read-only file system: OCI runtime error

This was fixed in https://github.com/opencontainers/runc/pull/3345 which was backported to runc 1.1 in https://github.com/opencontainers/runc/pull/3355, and the fix made its way to runc 1.1.1.

What runc version is used in the test?

**Comment by rhatdan** [comment]
@cevich would know for sure, but I think it is using the current released versions of runc in Ubuntu.

**Comment by cevich** [comment]
One of the failing Ubuntu 22.04 tasks reports:

```
Ubuntu 22.04 LTS \n \l
Kernel:  5.15.0-1013-gcp
Cgroups:  tmpfs
dpkg-query: warning: parsing file '/var/lib/dpkg/status' near line 1430 package 'containers-common':
 missing 'Maintainer' field
dpkg-query: no packages found matching golang
dpkg-query: no packages found matching cri-o-runc
conmon-2:2.1.2-0ubuntu22.04+obs5.1-amd64
containernetworking-plugins-0.9.1+ds1-1-amd64
containers-common-4:1-0ubuntu22.04+obs13.1-amd64
criu-3.17.1-1-amd64
crun-1.5-0ubuntu22.04+obs34.1-amd64
libseccomp2-2.5.3-2ubuntu2-amd64
podman-3:4.2.0~rc1-0ubuntu22.04+obs28.1-amd64
runc-1.1.0-0ubuntu1-amd64
skopeo-2:1.8.0-0ubuntu22.04+obs16.1-amd64
slirp4netns-1.2.0-0ubuntu22.04+obs4.1-amd64
```

Checking if there's an update available...There are 28 package updates available, none of them are runc.  We have an option to bring in a custom-built version, but generally the distro-provided packages would be preferred.  Is there a way we can tell if `runc-1.1.0-0ubuntu1-amd64` is the package with the needed backport?

cc: @lsm5 FYI for now.

**Comment by edsantiago** [comment]
This is my fault, #16240. Can you run `make docs`, then `git add` the changed file and `git commit --amend`?

File: docs/source/markdown/options/read-write-tmpfs.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+####> This option file is used in:
+####>   podman create, run
+####> If you edit this file, make sure your changes
+####> are applicable to all of those.
+#### **--read-write-tmpfs**
+
+If container is running in --read-only mode, then mount a read-write tmpfs on /run, /tmp, and /var/tmp. When false, Podman the entire container has no tmpfs that are read/write unless specified on the command line.  Podman mounts /dev, /dev/mqueue, /dev/pts, /dev/shm as read only. The default is *true*
```
Comment (edsantiago) [inline docs/source/markdown/options/read-write-tmpfs.md:7]: >When false, Podman the entire container

Does not parse. I think `Podman` is the part that needs to go, but I'm not sure what the intention was.

Did you deliberately un-underscore the directory names? Our guidelines say:

>Strings of characters or numbers can be highlighted with `backticks`. Paths of any kind **must** be highlighted.

(which suggests that backticks are preferred, underscores are ok, but leaving them unadorned is not).

File: test/system/030-run.bats [hunk 4]
```diff
@@ -832,6 +832,25 @@ EOF
...
+    fi
+    HOST=$(random_string 25)
+    run_podman 1 run --rm --read-only $IMAGE touch /foo
+    is "$output" "touch: /foo: Read-only file system" "Should fail with read only file system"
+
+    for rwdir in /run /tmp /var/tmp /dev /dev/shm; do
+        run_podman run --rm --read-only $IMAGE touch ${rwdir}/foo
+        run_podman run --rm --read-only --read-write-tmpfs=true $IMAGE touch ${rwdir}/foo
+	run_podman 1 run --rm --read-only --read-write-tmpfs=false $IMAGE touch ${rwdir}/foo
+	is "$output" "touch: ${rwdir}/foo: Read-only file system" "Should fail with ${rwdir}/foo read only file system"
```
Comment (edsantiago) [inline test/system/030-run.bats:850]: Inconsistent indentation: some of these are tabs, some are eight spaces.

File: test/system/030-run.bats [hunk 5]
```diff
@@ -832,6 +832,25 @@ EOF
     is "${lines[0]}" ".*${HOST}.*"
 }
 
+@test "podman run --read-write-tmpfs" {
+    # FIXME once ubuntu gets a newer crun
+    if is_ubuntu; then
+        if [ $(podman_runtime) == "crun" ]; then
+	   skip "The version of crun shipped by ubuntu is too old. Needs crun-1.4.2 or greater"
```
Comment (edsantiago) [inline test/system/030-run.bats:839]: I have a `find-obsolete-skips` script that I run periodically. Could you add "FIXME:" to this and the above `skip/Skips` ? That will bring these to my attention when I run the script next time.

File: test/system/030-run.bats [hunk 6]
```diff
@@ -832,6 +832,25 @@ EOF
...
+	fi
+    fi
+    HOST=$(random_string 25)
```
Comment (edsantiago) [inline test/system/030-run.bats:842]: leftover from copy/paste (hence misleading, because it's never used). Please remove.

File: cmd/podman/containers/create.go [hunk 7]
```diff
@@ -114,6 +114,11 @@ func commonFlags(cmd *cobra.Command) error {
 	if cmd.Flags().Changed("image-volume") {
 		cliVals.ImageVolume = cmd.Flag("image-volume").Value.String()
 	}
+
+	if cmd.Flags().Changed("read-write-tmpfs") && !cliVals.ReadOnly {
+		return errors.New("--read-write-tmpfs is only used if --read-only=true")
```
Comment (edsantiago) [inline cmd/podman/containers/create.go:119]: I think this would be clearer as "...is only **valid** if", or "...is only meaningful when..."

File: docs/source/markdown/podman-run.1.md.in [hunk 8]
```diff
@@ -477,7 +477,7 @@ tmpfs directories on _/run_ and _/tmp_.
 ```
 $ podman run --read-only -i -t fedora /bin/bash
 
-$ podman run --read-only --read-only-tmpfs=false --tmpfs /run -i -t fedora /bin/bash
+$ podman run --read-only --read-write-tmpfs=false --tmpfs /run -i -t fedora /bin/bash
```
Comment (edsantiago) [inline docs/source/markdown/podman-run.1.md.in:480]: (commenting on code that github does not show by default, so please click the uparrow-on-bar to see lines 473-475). Those lines read:
>The best way to handle this is to mount tmpfs directories on _/run_ and _/tmp_.

This looks like a holdover from a previous age: the implication is that `/run` and `/tmp` are also read-only. I haven't taken the time to go through history, but is it possible that the `read-only-tmpfs` option (the one that is being renamed here) was added after this was written, and that this used to be true but no longer is? Can you re-review lines 473-475 and see if they should be rewritten, maybe something like

>Read-only containers still mount `/run`, `/tmp`, `/var/tmp`, `/dev`, and `/dev/shm` as read-write tmpfs. Use `--read-write-tmpfs=false` to protect even `tmpfs` directories, and `--tmpfs=PATH` to explicitly limit the read-write tmpfs filesystems.

(Idea only. That reads horribly. My documentation-writing brain is not fully engaged yet).

**Review by edsantiago — CHANGES_REQUESTED** [review]
Lots of little nits, sorry for not catching them earlier.

File: docs/source/markdown/options/read-write-tmpfs.md [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+####>   podman create, run
+####> If you edit this file, make sure your changes
+####> are applicable to all of those.
+#### **--read-write-tmpfs**
+
+If container is running in --read-only mode, when true Podman mounts a
+read-write tmpfs on _/run_, _/tmp_, and _/var/tmp_. When false, Podman sets
+the entire container such that no tmpfs can be written to unless specified on
+the command line.  Podman mounts _/dev_, _/dev/mqueue_, _/dev/pts_, _/dev/shm_
+as read only. The default is *true*
```
Comment (edsantiago) [inline docs/source/markdown/options/read-write-tmpfs.md:11]: I find this hard to parse, and am really confused by the discrepancies between the first list of "these are tmpfs" and the second. I'm just going to hope that @TomSweeneyRedHat will offer his always-helpful input. Otherwise my (much less refined) thinking goes something like:
```markdown
Only valid in combination with **--read-only**.  Default: **true**.

Normally a container run with **read-only** will mount the following `tmpfs` filesystems read/write:
   [ complete enumeration of those filesystems, possibly with a reference to where in the code that list is found ]
With **--read-write-tmpfs=false**, even those filesystems are mounted read-only.
```

**Comment by rhatdan** [comment]
Replaced by https://github.com/containers/podman/pull/16545
